src/healthcheck.c

```diff
@@ -0,0 +1,645 @@
+
+#include "healthcheck.h"
+#include "utils.h"
+#include "ctr_logging.h"
+#include "parent_pipe_fd.h"
+#include "globals.h"
+#include "cli.h"
+#include "ctr_exit.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/wait.h>
+#include <sys/stat.h>
+#include <time.h>
+#include <signal.h>
+#include <pthread.h>
+#include <cJSON.h>
+
+/* Simple hash table implementation */
+struct hash_entry {
+    char *key;
+    void *value;
+    struct hash_entry *next;
+};
+
+struct hash_table {
+    struct hash_entry **buckets;
+    size_t size;
+    size_t count;
+};
+
+static unsigned int hash_string(const char *str) {
+    unsigned int hash = 5381;
+    int c;
+    while ((c = *str++)) {
+        hash = ((hash << 5) + hash) + c;
+    }
+    return hash;
+}
+
+struct hash_table *hash_table_new(size_t size) {
+    struct hash_table *ht = calloc(1, sizeof(struct hash_table));
+    if (!ht) {
+        return NULL;
+    }
+    
+    ht->buckets = calloc(size, sizeof(struct hash_entry*));
+    if (!ht->buckets) {
+        free(ht);
+        return NULL;
+    }
+    
+    ht->size = size;
+    return ht;
+}
+
+void hash_table_free(struct hash_table *ht) {
+    if (!ht) return;
+    
+    for (size_t i = 0; i < ht->size; i++) {
+        struct hash_entry *entry = ht->buckets[i];
+        while (entry) {
+            struct hash_entry *next = entry->next;
+            free(entry->key);
+            free(entry);
+            entry = next;
+        }
+    }
+    
+    free(ht->buckets);
+    free(ht);
+}
+
+void *hash_table_get(struct hash_table *ht, const char *key) {
+    if (!ht || !key) return NULL;
+    
+    unsigned int hash = hash_string(key) % ht->size;
+    struct hash_entry *entry = ht->buckets[hash];
+    
+    while (entry) {
+        if (strcmp(entry->key, key) == 0) {
+            return entry->value;
+        }
+        entry = entry->next;
+    }
+    
+    return NULL;
+}
+
+bool hash_table_put(struct hash_table *ht, const char *key, void *value) {
+    if (!ht || !key) return false;
+    
+    unsigned int hash = hash_string(key) % ht->size;
+    struct hash_entry *entry = ht->buckets[hash];
+    
+    /* Check if key already exists */
+    while (entry) {
+        if (strcmp(entry->key, key) == 0) {
+            entry->value = value;
+            return true;
+        }
+        entry = entry->next;
+    }
+    
+    /* Create new entry */
+    entry = malloc(sizeof(struct hash_entry));
+    if (!entry) return false;
+    
+    entry->key = strdup(key);
+    if (!entry->key) {
+        free(entry);
+        return false;
+    }
+    
+    entry->value = value;
+    entry->next = ht->buckets[hash];
+    ht->buckets[hash] = entry;
+    ht->count++;
+    
+    return true;
+}
+
+/* Global healthcheck timers hash table */
+struct hash_table *active_healthcheck_timers = NULL;
+
+/* Check if systemd is available and running */
+bool healthcheck_is_systemd_available(void) {
```

Comment:
Feels unnecessary, this should be in Podman. If we tell Conmon to do a healthcheck it should do it. If we don't, it's because systemd is available and we're preferring that.

Reply:
cleaned up. thanks